### PR TITLE
Quick description of µDraft

### PR DESCRIPTION
**Pull request by hanbzu**
This PR is a request for comments (RFC) style discussion for defining better the app before starting to code functionalities.

### CODE REVIEW

**Comment by hanbzu** [comment]
👋 Hello everyone!

Please have a look a the app description at https://github.com/hanbzu/xmas-one/tree/app-shaping-rfc (the README).

- How hard / easy is it to understand the idea of the app?
- What do you say, does it seem interesting as a project? Any ideas around this?

**Comment by hanbzu** [comment]
I added some examples of the data to be saved and the data to be shared https://github.com/hanbzu/xmas-one/blob/app-shaping-rfc/specs/data-examples.md

I know it's weird but those examples are, for now, the best way to explain what the app does.

As there's no backend database for persisting state, it lives in every browser, in an append log in local storage. In this append log you see all the versions of a text, your own versions and the ones others have suggested to you.

The app state is the reduction of this log + your current live editing. Whenever you click on a "share link" button, you solidify your live editing into a new immutable version that gets appended to your log (and it's also the content of the URL you are going to share with your friend).

This diagram tries to explain it functionally from the perspective of the interface:

image

_(this technique is explained in [Ryan Singer's Shape Up](https://basecamp.com/shapeup/1.3-chapter-04))_

**Comment by gedalos-dev** [comment]
A simple question about the length of urls ...

[What is the maximum length of a URL in different browsers?](https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers/62757970#62757970)


Then pass data with URL won't be a problem. 👍

**Comment by gedalos-dev** [comment]
In summary. 

Apart from suggest a new version of received text, you can add comments on it.

And of course you can only write some comments without alter received text.


This is correct ?

**Comment by gedalos-dev** [comment]
As a collaborator user I can follow the evolution of text ? 

Once I have been a collaborator... 
> I can get notified when owner **changes** text? or
> I can get notified only when owner **shares** the text?

**Comment by hanbzu** [comment]
> And of course you can only write some comments without alter received text.

You provide comments. And also you can provide a new text. But it's a new version to be added to the log/history.

So if I receive:

`Hi, I'm a triceratops`

I can change it to `Hello, I'm a triceratops`. But **I'm not really changing it**. When I edit, I'm creating a new version.

Let's see if this diagram helps:

image

**Comment by hanbzu** [comment]
> I can get notified when owner **changes** text? or
> I can get notified only when owner **shares** the text?

No. This is a very simple no backend involved version. The only way for information to travel between people's computers are links. The links contain new information.

If someone wants to give you access to a new draft version, they send you the link that contains it. You get notified in the email or chat program you use by seeing the person's message:

```
Hey gedalos, check out my text at https://microdraft.app/j2h342h3je7a9q34u32hjgdsd7tr2382k3hrkrew8rw3i4u923492u2edhkwedw8uwr8y34723ih2dkw83w783824u2kew8368r628iru2o3r9fefnhw3r08wr373
```

Then, you click that link and you see the text, AND you can start editing it. Once you're done you'll be able to send back your edits (suggestions) to your peer as a new version, which they'll know about the same way.

**Comment by gedalos-dev** [comment]
Following this example... https://github.com/hanbzu/xmas-one/pull/4#issuecomment-718812527

Then, if Ann still writing the text, (having accepted Bob's review), and wants to share it with Bob, (ex. a new final review of his text after a while), when Ann sends the new URL, Bobs receive the last version...

Can Bob see the previous revisions/records _"Pick another version"_? (as described in https://github.com/hanbzu/xmas-one/pull/4#issuecomment-718647719

**Comment by hanbzu** [comment]
@gedalos-dev exactly. Everything you get through a link gets appended to your local log.

And as the versions are labelled by UUIDs they are effectively saved for good. Additionally, versions refer (anchor) to another version. That means that your log is in fact a version tree. Any participant could choose any version to fork.

image

So, Ann and Bob could go lineally, back-and-forth sharing links:

```
1 (Ann) -> 2 (Bob) -> 3 (Ann) -> 4 (Bob)...
```

But Ann could choose not to continue from Bob's draft and make changes to her original draft instead:

```
1 (Ann) -> 2 (Bob)
1 (Ann) -> 1b (Ann) -> X (Bob)
```

**Comment by gedalos-dev** [comment]
### Possible Scenario ...

- [v1] author generates text and sends it to reviewer-1 and reviewer-2.
- [v2] author continues the text.
- [v3] reviewer-1 adds comments and a change and forwards it to the author.
- [v4] reviewer-2 makes a change and forwards it to the author.
- [v5] author receives version of reviewer-1 and adapts the text, adds a comment and sends it to reviewer-1.
- [v6] author receives version of reviewer-2 and accepts its change.
- [v7] reviewer-1 based on the new comment makes a change and forwards it to the author.
- [v8] author receives version of reviewer-1 and discard change from reviewer-1.

> Then dependency versions...

```
       --> v2
      |
v1 --> --> v3 --> v5 --> v7 --> v8 --> v7
      |
       --> v4 --> v6
```

**I may miss something** what about v2, v7 and v6 ?

**Comment by hanbzu** [comment]
> **I may miss something** what about v2, v7 and v6 ?

First of all, thanks for playing this game.

You are right, as soon as you include more than one person it becomes a bit complex to keep track of versions. The issue is **you may want to say your new version derives from more than one version** (in your example v5 could be the result of _merging_ v2 and v3).

Allowing for anchoring a version to more than one version...

```js
{ anchor: ["v2", "v3"],
  uuid: "v5",
  shared: "2020-10-29T11:12:00Z",
  author: "puffin",
  text: "..."
}
```

...maybe helps with keeping track of versions. You you either, "dismiss" them or you anchor your new version to them (~accept, merge).

@gedalos-dev does this make sense?

**Comment by gedalos-dev** [comment]
```
       --> v2
      |
v1 --> --> v3 --> v5 --> v7 --> v8 --> v7
      |
       --> v4 --> v6
```

..._`(in your example v5 could be the result of merging v2 and v3).`_

...then... in time(t):

t(1): author receives version of reviewer-1 (v1 -> v3) and adapts the text, adds a comment and sends it to reviewer-1.

#### case a: v1 vs v3: (is it possible to compare with old versions?  v1 is an old version due to v2). 

t(2a): author compares v1 and v3 and after changes then he creates a v13.
t(3a): author compares v2 and v13 and after changes then he creates a v5.

#### case b: v2 vs v3: 

t(2b): author compares v2 and v3 and after changes then he creates a v5.


In my head I can not ignore the git mechanism, and that is influencing me. This is how to play with git and its branches and what merge method to use. Maybe i'm not right.

_I like to play :smile:_

**Comment by hanbzu** [comment]
Yes, both those scenarios are possible.

> is it possible to compare with old versions? v1 is an old version due to v2

Yes, in my mid, in the diff view, you can toggle any of the two versions you are comparing. So, if you access the website through a link that someone sent you, by default the diff view shows that version in the right and the anchor(s) (parents) of that version in the left. **But you have a dropdown or similar at the top of the diff view where you can decide any sort of comparison.**

So, yes, git-like workflows are sort of possible with one caveat: **whenever someone sends you a link you get ONE version, not all the versions they made till that version.** Example:

Ann shares v1 with Bob. But Bob, creates v2, then v3, and then decides to send changes to Ann
```
t    0     1     2     3     4
Ann  v1                      v3
Bob        v1 -> v2 -> v3
```
Then the version tree will look:
- `v1 -> v3` for Ann,
- `v1 -> v2 -> v3` for Bob

In other words, Ann will not see v2. Which is not crucial **as long as you are aware that others cannot see any version that you don't share with them and they get to open** (it needs to be clear, like highlighted somehow in the UI, what you are sending when you "share link").

**Comment by hanbzu** [comment]
I have updated the diagram to take into account the things we've talked about:

image
